### ISSUE TITLE
Build new network above the 128d representational layer 

### ISSUE THREAD

**Issue by wkcw** [davidsandberg/facenet#82]
I'm new to deep learning and tf, so the 2 questions may be basic
1.  Is  inception_resnet_v1.py the model graph script? I was a bit confused by the files, but I think I should modify the model graph here to build new network
2.  I want to build another network on the basis of the 128d fc layer(according to the paper)
, which means the pre-trained inception v1 net will remain unchanged during the training
of new layers.
but I don't know which file should I modify to achieve the goal.
plus, I did not see the 128d last layer in inception_resnet_v1.py, could it be that I misundertand the code?

My sincere gratitude!

**Comment by davidsandberg** [comment]
I think you can have a look at facenet_train.py and the parameter pretrained_model. I haven't used it for a while so I'm not 100% sure that it works but it should give you some idea of how to continue training of a model and how to block gradient updates for the rest of the model except for the final layer.
The inception_resnet_v1.py is the module that creates the inception resnet v1 but the final layer depends on the task (classification or triplet loss) so that is created by the training module, i.e. facenet_train.py or facenet_train_classifier.py. 
Good luck!

**Comment by wkcw** [comment]
@davidsandberg 
Thank you for your response!
One more thing I want to figure out 
I've downloaded the pretrained model-20160506.ckpt-500000 and found it contains a tensor named 'embeddings'.  I used to consider 'embeddings' as the last layer of this model file(which should be 1x128 according to the paper) and I thought I should start my model from here.

But after checking it with 
```
embeddings = tf.get_default_graph().get_tensor_by_name("embeddings:0")
print(tf.Tensor.get_shape(embeddings))
```
the output is 
`(?, 1792)`
it seems that 'embbedings' is 1792d

Briefly, I thought the model is trained with facenet_train.py and drew the conclusion that 'embeddings' is 128d, for the codes
```
pre_embeddings = slim.fully_connected(prelogits, 128, activation_fn=None, scope='Embeddings',
reuse=False)

# Split example embeddings into anchor, positive and negative and calculate triplet loss
embeddings = tf.nn.l2_normalize(pre_embeddings, 1, 1e-10, name='embeddings')
```
So if 'embeddings' is not 128d, which script trains the pretrained model?
Or did I made a mistake in checking the shape of embeddings?

Sorry for the lengthy question.

**Comment by ugtony** [comment]
I had the same question before. So I manually added a fully connected layer at the end of inception_resnet_v1.py to and ran some experiments. It worked. The performance is slightly better when it is 512d and worse when 128d. 

My next question is why the L2-Nomalization op is put only for evaluation and not for traning (before center loss).  I am working on this experiment now.

**Comment by davidsandberg** [comment]
Yes, @ugtony is very correct. The dimension of the embedding when running facenet_train_classifier.py is set by the model that is used, which is the inception_resnet_v1 model for most of the later models. But the model model-20160506.ckpt-500000 is trained using facenet_train.py and doesn't perform very well.
Also, TF r0.10 doesn't work very well with this repo. It was recently updated to work well with TF r0.12 so from now on I don't think it works very well with earlier releases. Not sure though...

**Comment by sarthakahuja11** [comment]
@ugtony  and @wkcw I am working on  a similar problem, and am planning to add new layers to the pre-trained model. I just had to switch from keras to tensorflow as the model is in tf. I have just started with tf, so could you guide me how to go about adding new layers after restoring the model? Any link or code spinnet would do.

Thank you!

**Comment by saisubramani** [comment]
shall i able to get the 1*128 embeddings of each images  .?
how can i achive this?
so that i can store it as numpy file for my research.
